csharp/src/Ice/IceSSL/TransceiverI.cs

```diff
@@ -512,278 +449,117 @@ private void finishAuthenticate()
         }
     }
 
-    private X509Certificate selectCertificate(
-        object sender,
-        string targetHost,
-        X509CertificateCollection certs,
-        X509Certificate remoteCertificate,
-        string[] acceptableIssuers)
-    {
-        if (certs == null || certs.Count == 0)
-        {
-            return null;
-        }
-        else if (certs.Count == 1)
-        {
-            return certs[0];
-        }
-
-        // Use the first certificate that match the acceptable issuers.
-        if (acceptableIssuers != null && acceptableIssuers.Length > 0)
-        {
-            foreach (X509Certificate certificate in certs)
-            {
-                if (Array.IndexOf(acceptableIssuers, certificate.Issuer) != -1)
-                {
-                    return certificate;
-                }
-            }
-        }
-        return certs[0];
-    }
-
     private bool validationCallback(
         object sender,
         X509Certificate certificate,
         X509Chain chainEngine,
         SslPolicyErrors policyErrors)
     {
         using var chain = new X509Chain(_instance.engine().useMachineContext());
-        try
+
+        bool verified = false;
+        int errors = (int)policyErrors;
+        int traceLevel = _instance.securityTraceLevel();
+        string traceCategory = _instance.securityTraceCategory();
+        Ice.Logger logger = _instance.logger();
+        bool checkCertName = _instance.engine().getCheckCertName();
+        string message = "";
+
+        X509Certificate2Collection caCerts = _instance.engine().caCerts();
+        X509ChainStatus[] chainStatus = [];
+        if (caCerts is null)
+        {
+            verified = policyErrors == SslPolicyErrors.None;
+            chainStatus = chainEngine.ChainStatus;
+        }
+        else
         {
-            if (_instance.checkCRL() == 0)
+            chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+            chain.ChainPolicy.DisableCertificateDownloads = true;
+            chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+            foreach (X509Certificate2 cert in caCerts)
```

Comment:
Here we default to using the system CAs unless the users provide a custom CA. If the user provides the CA we build a new chain setting `X509ChainTrustMode.CustomRootTrust`.